Pulumi Deployments Relaunch (#2591)
pulumi/pulumi-hugo: themes/default/content/docs/intro/deployments/_index.md

---
title_tag: "Pulumi Deployments"
title: "Pulumi Deployments"
meta_desc: Pulumi Deployments is the fast way for individuals, teams, enterprises, and platforms to go from code to cloud.
menu:
  intro:
    identifier: deployments
    weight: 6

aliases: ["/docs/intro/pulumi-service/deployments/"]
---

Pulumi Deployments is a fully managed platform for managing infrastructure at scale. Whether you are just getting started in the cloud and want the easiest path to set up secure CI/CD, or are a platform engineer building a self-service platform to support thousands of engineers, Pulumi Deployments is the fastest way to go from code to cloud. We provide the building blocks that enable you to scale your cloud footprint without scaling your headcount, with tools that are designed to power public and private cloud platforms, APIs, and self-service infrastructure portals.

It provides:

> - **Compute** - run infrastructure deployments and workloads (ie `pulumi up`) on fully managed, secure, and elastic compute

Reviewer: ```suggestion
- **Compute** - run infrastructure deployments and workloads (i.e. `pulumi up`) on fully managed, secure, and elastic compute.
```

> - **Configuration** - A simple, centralized format for specifying everything you need to deploy your infrastructure at the stack level. Source code, cloud credentials, OpenID Connect (OIDC), environment variables, build prerequisites, and dependencies are bundled in an atomic unit

Reviewer: ```suggestion
- **Configuration** - A simple, centralized format for specifying everything you need to deploy your infrastructure at the stack level. Source code, cloud credentials, OpenID Connect (OIDC), environment variables, build prerequisites, and dependencies are bundled in an atomic unit.
```

> - **Composition** - Trigger deployments via our REST API, click to deploy in the Pulumi Service UI, Git Push to Deploy, Remote Automation API, and other programmatic building blocks that you can compose to build custom cloud automation

Reviewer: ```suggestion
- **Composition** - Trigger deployments via our REST API, click to deploy in the Pulumi Service UI, Git Push to Deploy, Remote Automation API, and other programmatic building blocks that you can compose to build custom cloud automation.
```

> While Pulumi Deployments has some overlap in functionality offered by a traditional CI/CD system (such as GitHub Actions) it is far more flexible and does not have the constraints and limitations of a CI/CD system. When you configure your infrastructure for Deployments, you get to choose what triggers will cause your stack to update. You may choose a `git push` based workflow, as you get with many CI/CD systems, or you may trigger deployments via the REST API, or you might choose multiple triggers.

Reviewer: ```suggestion
While Pulumi Deployments has some overlap in functionality offered by a traditional CI/CD system (such as GitHub Actions) it is far more flexible and does not have the constraints and limitations of a CI/CD system. When you configure your infrastructure for Deployments, you get to choose what triggers will cause your stack to update. You may choose a `git push` based workflow, as you get with many CI/CD systems, or trigger deployments via the REST API, or use a combination of triggers.
```

> While Pulumi Deployments has some overlap in functionality offered by a traditional CI/CD system (such as GitHub Actions) it is far more flexible and does not have the constraints and limitations of a CI/CD system. When you configure your infrastructure for Deployments, you get to choose what triggers will cause your stack to update. You may choose a `git push` based workflow, as you get with many CI/CD systems, or you may trigger deployments via the REST API, or you might choose multiple triggers.

Reviewer 2: I'm not sure I agree with this and think it's dangerous territory. I don't think we can say we're more flexible than say, github actions, but we can say we are better at infrastructure deployment.

No concrete ideas for improvement off the cuff, but I'd lean towards not even bringing up CI/CD unless we think we can't avoid it.

Author: We are more flexible than github actions. We have a REST API to trigger a workflow run. That API can be called from within lambda, ECS, temporal, or any other higher level software system you want to build. 

I think we have to bring up CI/CD. Customers very often ask us how this compares to github actions and why they would use one over the other. 

Open to adjustments here if you have ideas.

Author: Notes from Lee: 

Maybe more choice, more modular, more composable. Because this is a first class experience designed from the ground up for infrastructure. 

YOu have three parts:

- trigger
- configuration
- program

choose your own adventure

Author: The word flexibility may be a trigger for some people.

![Pulumi Deployments Platform Architecture](deployments.png)

> ## Deployment Settings

Reviewer 2: I think these sections are so clear and concise it's worth calling it out, ie:

> Each deployment has 3 parts. The settings, the trigger and the program

Each stack in your organization can be configured with deployment settings, a manifest that describes everything needed to run an update. Deployment settings turns a stack into an atomic, deployable unit of infrastructure that self-describes all of its requirements and dependencies. That includes:

- Source Control - the Pulumi program and stack configuration
- Deployment Prerequisites - custom docker images, build steps, npm package resolution, and acquisition of necessary tools like `ssh`, and environment variables
- Cloud Auth - IAM roles, OIDC configuration for temporary credentials, and any other identity-related config required to deploy into your cloud of choice

> These settings can be configured through the Pulumi Service Console, and via a REST API. The console interface is simple and lets you configure settings in just a few minutes, and the REST API allows you to build tools on top of Pulumi Deployments such as internal dev portals and service catalogs that help internal users kickstart their infrastructure journey with a secure, best practice workflow.

Reviewer: ```suggestion
These settings can be configured through the Pulumi Service Console, or via a REST API. The console interface is simple and lets you configure settings in just a few minutes, while the REST API allows you to build tools on top of Pulumi Deployments such as internal dev portals and service catalogs that help internal users kickstart their infrastructure journey with a secure, best practice workflow.
```

Once configured, every deployment will inherit stack settings. A one-line POST payload can trigger a new deployment run, without having to restate any configuration or cloud credentials. This means that privileged users can configure deployments, and less privileged users can subsequently trigger updates without having direct access to sensitive credentials and data on their local machines.

## Deployment Triggers

Pulumi Deployments offers several triggers that you can compose to build higher-level infrastructure automation. Because each trigger inherits deployment settings, you only have to configure a stack once.

> - Deployments REST API - an endpoint available at api.pulumi.com that can trigger a variety of pulumi (`up/destroy/refresh/preview`) operations. Calls to this trigger can override individual settings properties to customize deployment behavior (ie - deploy a different branch, or modify a particular environment variable)

Reviewer: ```suggestion
- Deployments REST API - an endpoint available at api.pulumi.com that can trigger a variety of pulumi (`up/destroy/refresh/preview`) operations. Calls to this trigger can override individual settings to customize deployment behavior (i.e. - deploy a different branch, or override a particular environment variable).
```

> - `git push` to Deploy - Deep integration with the Pulumi GitHub App that enables automatic previews on every pull request, and an automated update whenever code is pushed to the target branch

Reviewer: ```suggestion
- `git push` to Deploy - Deep integration with the Pulumi GitHub App that enables automatic previews on every pull request, and an automated update whenever code is pushed to the target branch.
```

> - Click to Deploy - trigger a deployment from the Pulumi Service Console with the push of a button. Very useful for day-to-day operational activities such as debugging a stuck stack, or rectifying drift via a refresh, without having to pull a stack and its accompanying source code onto your machine.

Reviewer: ```suggestion
- Click to Deploy - trigger a deployment from the Pulumi Service Console with the push of a button. Very useful for day-to-day operational activities such as debugging a stuck stack or rectifying drift via a refresh without having to pull a stack and its accompanying source code onto your machine.
```

- Remote Automation API - Your favorite strongly typed Pulumi orchestration tool, Automation API, lets you offload update workloads into the cloud. Write deployment scripts that update thousands of environments simultaneously without worrying about compute.

Triggers can be used together, or not at all. Some users like to use the REST API to trigger deployments from GitHub Actions and disable the built-in `git push` integration. Doing so gives them the flexibility to compose deployment settings, the REST API, and click to deploy into useful platform workflows outside of CI/CD.

> Deployment settings plus triggers are the key building blocks of the deployment platforms. With them, you can build highly customized infrastructure automation and developer platforms such as service catalogs, drift detection, temporary infrastructure workflows for cost savings, infrastructure REST APIs, and more. These scenarios and more are discussed below with demos and source code.

Reviewer: ```suggestion
Deployment settings plus triggers are the key building blocks of the deployment platforms. With them, you can build highly customized infrastructure automation and developer platforms such as service catalogs, drift detection, temporary infrastructure workflows for cost savings, infrastructure REST APIs, and more. These scenarios and more are discussed further with demos and source code.
```
